keylime/cert_utils.py

```diff
@@ -1,13 +1,90 @@
-from cryptography.hazmat.primitives.serialization import load_der_public_key
+import io
+import sys
+
+from cryptography import exceptions as crypto_exceptions
+from cryptography import x509
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives.asymmetric import ec, padding
+from cryptography.hazmat.primitives.asymmetric.ec import EllipticCurvePublicKey
+from cryptography.hazmat.primitives.asymmetric.rsa import RSAPublicKey
 from pyasn1.codec.der import decoder, encoder
-from pyasn1_modules import rfc2459
+from pyasn1_modules import pem, rfc2459
 
+from keylime import config, keylime_logging, tpm_ek_ca
 
 # Issue #944 -- python-cryptography won't parse malformed certs,
 # such as some Nuvoton ones we have encountered in the field.
 # Unfortunately, we still have to deal with such certs anyway.
-# Let's read the EK cert with pyasn1 instead of python-cryptography.
-def read_x509_der_cert_pubkey(der_cert_data):
-    """Returns the public key of a DER-encoded X.509 certificate"""
-    der509 = decoder.decode(der_cert_data, asn1Spec=rfc2459.Certificate())[0]
-    return load_der_public_key(encoder.encode(der509["tbsCertificate"]["subjectPublicKeyInfo"]))
+
+# Here we provide some helpers that use pyasn1 to parse the certificates
+# when parsing them with python-cryptography fails, and in this case, we
+# try to read the parsed certificate again into python-cryptograhy.
+
+logger = keylime_logging.init_logging("cert_utils")
+
+
+def x509_der_cert(der_cert_data):
+    if isinstance(der_cert_data, str):
+        der_cert_data = der_cert_data.encode("utf-8")
+    try:
+        cert = cert = x509.load_der_x509_certificate(data=der_cert_data, backend=default_backend())
+        return cert
+    except Exception:
+        pyasn1_cert = decoder.decode(der_cert_data, asn1Spec=rfc2459.Certificate())[0]
+        return x509.load_der_x509_certificate(data=encoder.encode(pyasn1_cert), backend=default_backend())
+
+
+def x509_pem_cert(pem_cert_data):
+    if isinstance(pem_cert_data, (bytes, bytearray)):
+        pem_cert_data = pem_cert_data.decode("utf-8")
```

Comment:
maybe it should be documented what types of data can be passed into this function. Ideally it would just be one datatype that 'pem_cert_data' could be...

Reply:
Yeah, I tend to agree with this. I changed `x509_pem_cert()` to accept `str` and `x509_der_cert()` to accept `bytes`, for now.